ports/zephyr/machine_uart.c

```diff
@@ -60,23 +77,95 @@ static void mp_machine_uart_print(const mp_print_t *print, mp_obj_t self_in, mp_
+        { MP_QSTR_flow, MP_ARG_KW_ONLY | MP_ARG_INT, {.u_int = 0} },
     };
+
     mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
     mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
 
     self->timeout = args[ARG_timeout].u_int;
     self->timeout_char = args[ARG_timeout_char].u_int;
+
+    uint8_t data_bits;
+    if (args[ARG_bits].u_int == 5) {
+        data_bits = UART_CFG_DATA_BITS_5;
+    } else if (args[ARG_bits].u_int == 6) {
+        data_bits = UART_CFG_DATA_BITS_6;
+    } else if (args[ARG_bits].u_int == 7) {
+        data_bits = UART_CFG_DATA_BITS_7;
+    } else if (args[ARG_bits].u_int == 8) {
+        data_bits = UART_CFG_DATA_BITS_8;
+    } else if (args[ARG_bits].u_int == 9) {
+        data_bits = UART_CFG_DATA_BITS_9;
+    } else {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("invalid data bits"));
+    }
+
+    uint8_t parity;
+    if (args[ARG_parity].u_obj == mp_const_none) {
+        parity = UART_CFG_PARITY_NONE;
+    } else if (mp_obj_get_int(args[ARG_parity].u_obj) == 0) {
+        parity = UART_CFG_PARITY_EVEN;
+    } else if (mp_obj_get_int(args[ARG_parity].u_obj) == 1) {
+        parity = UART_CFG_PARITY_ODD;
+    } else {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("invalid parity"));
+    }
+
+    uint8_t stop_bits;
+    if (args[ARG_stop].u_int == 1) {
+        stop_bits = UART_CFG_STOP_BITS_1;
+    } else if (args[ARG_stop].u_int == 2) {
+        data_bits = UART_CFG_STOP_BITS_2;
+    } else {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("invalid stop bits"));
+    }
+
+    uint8_t flow_ctrl;
+    if (args[ARG_flow].u_int == 0) {
+        flow_ctrl = UART_CFG_FLOW_CTRL_NONE;
+    } else if (args[ARG_flow].u_int == (MACHINE_UART_RTS | MACHINE_UART_CTS)) {
+        flow_ctrl = UART_CFG_FLOW_CTRL_RTS_CTS;
+    } else {
+        mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("invalid flow control"));
+    }
+
+    const struct uart_config cfg = {
+        .baudrate = args[ARG_baudrate].u_int,
+        .parity = parity,
+        .stop_bits = args[ARG_stop].u_int,
+        .data_bits = data_bits,
+        .flow_ctrl = flow_ctrl
+    };
+
+    int ret = uart_configure(self->dev, &cfg);
+    if (ret < 0) {
+        mp_raise_OSError(-ret);
+    }
+
+    ringbuf_alloc(&self->tx_ringbuffer, args[ARG_txbuf].u_int);
+    ringbuf_alloc(&self->rx_ringbuffer, args[ARG_rxbuf].u_int);
+
+    uart_irq_callback_user_data_set(self->dev, uart_interrupt_handler, (void *)self);
+    uart_irq_rx_enable(self->dev);
 }
 
 static mp_obj_t mp_machine_uart_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
     mp_arg_check_num(n_args, n_kw, 1, MP_OBJ_FUN_ARGS_MAX, true);
 
-    machine_uart_obj_t *self = mp_obj_malloc(machine_uart_obj_t, &machine_uart_type);
+    machine_uart_obj_t *self = mp_obj_malloc_with_finaliser(machine_uart_obj_t, &machine_uart_type);
     self->dev = zephyr_device_find(args[0]);
```

Comment:
If `zephyr_device_find()` raises an exception (eg unknown UART id) then `self->dev` has an undefined value, and when the GC reclaims this object it tries to disable IRQs on an undefined object, leading to a crash.

Eg on frdm_k64f, do `machine.UART(1)` and then soft reset, and it'll lock up.

Suggested fix:
```c
const struct device *dev = zephyr_device_find(args[0]);

machine_uart_obj_t *self = mp_obj_malloc_with_finaliser(machine_uart_obj_t, &machine_uart_type);
self->dev = dev;
self->tx_complete = true;
```
(That follows how `machine_i2c.c`, `machine_spi.c` etc do it.)

Reply:
Oh I see... Thanks for testing such scenario. I'll fix that.

Reply:
I only found it by accident, running a test that was meant for a different board!